testsuite/tests/typing-local/local.ml

```diff
@@ -320,15 +320,57 @@ val apply2 : int -> local_ 'a -> int -> int = <fun>
 |}]
 let apply3 x = g x x x
 [%%expect{|
-Line 1, characters 15-22:
+Line 1, characters 19-20:
 1 | let apply3 x = g x x x
-                   ^^^^^^^
+                       ^
+Error: This application involving locals is complete after this argument,
+       but extra arguments were provided
+  Hint: Try wrapping the application in parentheses up to this argument
+|}]
+let apply3_wrapped x = (g x x) x
+[%%expect{|
+Line 1, characters 23-32:
+1 | let apply3_wrapped x = (g x x) x
+                           ^^^^^^^^^
 Error: This local value escapes its region
   Hint: Cannot return local value without an explicit "local_" annotation
 |}]
 let apply4 x = g x x x x
 [%%expect{|
-val apply4 : int -> int = <fun>
+Line 1, characters 19-20:
+1 | let apply4 x = g x x x x
+                       ^
+Error: This application involving locals is complete after this argument,
+       but extra arguments were provided
+  Hint: Try wrapping the application in parentheses up to this argument
```

Comment:
I think a little time iterating on the error message may be time well spent. Here is my stab:

```suggestion
Line 1, characters 19-20:
1 | let apply4 x = g x x x x
                   ^^^^^
Error: The function application this message points to is complete,
       but extra arguments were provided.
       This is not allowed with local arguments.
  Hint: Try wrapping this application in parentheses
```

The big change here is to label the entire application, not just the last argument. (Not sure how hard that is in practice, but hopefully not too bad.) It took me a little while to understand "this argument".

I'm still not thrilled with my message, but maybe it inspires you do continue to improve.

Reply:
Yes, that does look better. It's mildly annoying to implement because we don't have that location to hand (there is no corresponding parsetree node), but the improvement seems worth it.

Reply:
This was easier than I expected! Better error message now pushed.